cmd/crank/beta/validate/manager.go

```diff
@@ -257,15 +283,152 @@ func (m *Manager) cacheDependencies() error {
 			return errors.Wrapf(err, "cannot download package %s", image)
 		}
 
-		schemas, _, err := extractPackageContent(*layer)
+		schemas, meta, err := extractPackageContent(*layer)
 		if err != nil {
 			return errors.Wrapf(err, "cannot extract package file and meta")
 		}
 
-		if err := m.cache.Store(schemas, path); err != nil {
+		// Cache schemas with meta
+		if err := m.cache.Store(append([][]byte{meta}, schemas...), path); err != nil {
 			return errors.Wrapf(err, "cannot store base layer")
 		}
 	}
 
 	return nil
 }
+
+func getDependencies(pkg interface{}) ([]metav1.Dependency, error) {
+	switch v := pkg.(type) {
+	case *metav1.Configuration:
+		return v.GetDependencies(), nil
+	case *metav1.Provider:
+		return v.GetDependencies(), nil
+	case *metav1.Function:
+		return v.GetDependencies(), nil
+	default:
+		return nil, errors.New("unknown package type")
+	}
+}
+
+func findPackageYamlType(meta []byte) (interface{}, error) {
+	// Define a list of possible types
+	candidates := []interface{}{
+		&metav1.Configuration{},
+		&metav1.Provider{},
+		&metav1.Function{},
+	}
+
+	// Try to unmarshal into each type
+	for _, candidate := range candidates {
+		if err := yaml.Unmarshal(meta, candidate); err == nil {
+			// If successful, return the candidate
+			return candidate, nil
+		}
+	}
+
+	return nil, errors.New("cannot unmarshal dependency YAML")
+}
+
+// getAndExtractPackageWithType gets the package image from cache or remote
+// and returns the package with its type (provider/function/configuration).
+func (m *Manager) getAndExtractPackageWithType(image string) (interface{}, error) {
```

Comment:
maybe split this function into two: one for reading it from cache and one for downloading and storing to focus on primary responsibilities

Reply:
I divided this function's content to 2 different function. One downloads the package from remote source and other tries to get the package from the cache. You can see the changes at https://github.com/crossplane/crossplane/pull/5874/commits/ebe648363be45bbd9693e2d31a37b9a0486a4ce0